pkg/reconciler/semantic.go

```diff
@@ -0,0 +1,99 @@
+/*
+Copyright (c) 2020 TriggerMesh, Inc
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package reconciler
+
+import (
+	appsv1 "k8s.io/api/apps/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/conversion"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/labels"
+)
+
+// Semantic can do semantic deep equality checks for Kubernetes API objects.
+//
+// For a given comparison function
+//
+//   comp(a, b interface{})
+//
+// 'a' should always be the desired state, and 'b' the current state for
+// DeepDerivative comparisons to work as expected.
+var Semantic = conversion.EqualitiesOrDie(
+	deploymentEqual,
+)
```

Comment:
Diff detection between the current and desired state of an API object uses deep derivative comparisons. In other words, when comparing object `A` to object `B`, only the fields that are actually set in `A` are compared to `B` (including individual elements of slices and maps). This provides a generic way to compare common API objects which works well in most scenarios, and simplifies the comparison code quite a lot.
Besides, it is resilient to external changes. For example, a binding may add a volume to an adapter's Deployment. With a deep comparison, the equality check would be false after such event, and the reconciler would always try to remove the new volume. With deep derivatives, the comparison remains true, as long as all the volumes from the desired state are identical to the ones in the current state.